apps/cms.c

```diff
@@ -721,15 +721,15 @@ int cms_main(int argc, char **argv)
         goto end;
 
     if (digestname != NULL) {
-        if (!opt_md(digestname, &sign_md))
+        if (!opt_md(app_get0_libctx(), digestname, app_get0_propq(), &sign_md))
```

Comment:
Since you now call `opt_md()` always with `app_get0_libctx()`,
why not simply make the libctx implicit, moving  `app_get0_libctx()` into `opt_md()`?

Reply:
... because the test programs use opt.c but not all the rest of libapps, and must certainly not apps.c, as that comes with contradicting baggage.  #15342 is an effort to clean up libapps and make it easier to use for the test programs, and would essentially be required to do what you propose.  However, we're too close to beta1 for that effort, it's been rightfully postponed to post-3.0.0.

So this is a functional compromise that makes things work without having to refactor all of libapps.

Reply:
I don't understand this, since `opt_md()` is use only in rather few places:
```
fgrep 'opt_md(' -r apps test
```
yields
```
apps/dgst.c:        if (!opt_md(digestname, &md))
apps/ocsp.c:            if (!opt_md(opt_arg(), &resp_certid_md))
apps/ocsp.c:            if (!opt_md(opt_unknown(), &cert_id_md))
apps/ocsp.c:        if (!opt_md(respdigname, &rsign_md))
apps/smime.c:        if (!opt_md(digestname, &sign_md))
apps/include/opt.h:int opt_md(const char *name, EVP_MD **mdp);
apps/lib/opt.c:int opt_md(const char *name, EVP_MD **mdp)
apps/crl.c:        if (!opt_md(digestname, &digest))
apps/pkcs12.c:            if (!opt_md(macalg, &macmd))
apps/cms.c:        if (!opt_md(digestname, &sign_md))
apps/ts.c:        if (!opt_md(digestname, &md))
apps/storeutl.c:        if (!opt_md(digestname, &digest))
apps/enc.c:        if (!opt_md(digestname, &dgst))
```
and AFICS by this PR all uses take `app_get0_libctx()` (and not NULL) as argument.

Reply:
If you move `app_get0_libctx()` into `opt_md()`, but not into `opt_md_silent()`, 
you still have the freedom to use NULL libctx via `opt_md_silent()`.

Reply:
That's not where the problem is.  The problem is that `app_get0_libctx()` and `app_get0_propq()` are defined in `apps/lib/apps.c`.  So, test programs would therefore need to have that as a source as well.  However, that sucks in a number of other symbols (functions) that test programs define their own variant of, and we get a number of symbol clashes when linking.

Of course, we could move `app_get0_libctx()` and `app_get0_propq()` to a separate source file.  However, they in turn depend on things like bio_err, bio_out and whatnot.  It rapidly becomes a mess to deal with.

Look, *you're welcome to try and see for yourself*, but be warned that it's not a small effort to deal with.
I'm choosing a route that *works* and is the least amount of effort.

Reply:
Ah, now I understand the problem - thanks.

Reply:
Still it would be possible to avoid adding `app_get0_libctx()` and `app_get0_propq()` to each and every call of `opt_md()` and `opt_cipher()` in the apps: by introducing intermediate `app_md()` and `app_cipher()` that take over this tedious job.

Reply:
... and where would you place those?

Reply:
In `apps.h`, for instance like this:
```
#define apps_md(n, d) opt_md(app_get0_libctx(), n, app_get0_propq(), d)
#define apps_cipher(n, c) opt_cipher(app_get0_libctx(), n, app_get0_propq(), c)
```

Reply:
I think that defining the `app_md` macros to just add arguments is wrong for a couple of reasons:
- @levitte already did the work to add them to the calls
- Since `opt_` is intended primarily as "option parsing for the application," creating a new prefix `app_` that wraps the `opt_` function is very confusing.

In other words, this adds more work (undo the changes already done), and add new macros to avoid the changes that result in more confusing code.  You could mitigate the confusion by having  big block comment explaining why `app_md` exists, but that still doesn't change the fact that things work now.